File: Core-DSA-Optimized/Graphs/Tarjans-SCC/tarjan_scc.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>
#include <cassert>

class TarjanSCC {
    int id = 0;
    std::vector<int> ids, low;
    std::vector<bool> onStack;
    std::stack<int> st;
    std::vector<std::vector<int>> sccs;

    void dfs(int at, const std::vector<std::vector<int>>& graph) {
        st.push(at);
        onStack[at] = true;
        ids[at] = low[at] = id++;

        for (int to : graph[at]) {
            if (ids[to] == -1) dfs(to, graph);
            if (onStack[to]) low[at] = std::min(low[at], low[to]);
        }

        if (ids[at] == low[at]) {
            std::vector<int> scc;
            int node = -1;
            while (node != at) {
                node = st.top(); st.pop();
                onStack[node] = false;
                scc.push_back(node);
            }
            sccs.push_back(scc);
        }
    }

public:
    std::vector<std::vector<int>> findSCCs(int n, const std::vector<std::vector<int>>& graph) {
        ids.assign(n, -1);
        low.assign(n, 0);
        onStack.assign(n, false);
        
        for (int i = 0; i < n; i++) {
            if (ids[i] == -1) dfs(i, graph);
        }
        return sccs;
    }
};

// --- CI/CD Automated Test ---
int main() {
    std::vector<std::vector<int>> graph = {{1}, {2}, {0}, {4}, {}};
    TarjanSCC tarjan;
    auto sccs = tarjan.findSCCs(5, graph);
    
    assert(sccs.size() == 3);
    std::cout << "C++ Tarjan SCC Test Passed!\n";
    return 0;
}

File: Core-DSA-Optimized/Graphs/Tarjans-SCC/tarjan_scc.py
class TarjanSCC:
    def __init__(self, n, graph):
        self.n = n
        self.graph = graph
        self.id = 0
        self.ids = [-1] * n
        self.low = [0] * n
        self.on_stack = [False] * n
        self.stack = []
        self.sccs = []

    def _dfs(self, at):
        self.stack.append(at)
        self.on_stack[at] = True
        self.ids[at] = self.low[at] = self.id
        self.id += 1

        for to in self.graph[at]:
            if self.ids[to] == -1:
                self._dfs(to)
            if self.on_stack[to]:
                self.low[at] = min(self.low[at], self.low[to])

        if self.ids[at] == self.low[at]:
            scc = []
            while True:
                node = self.stack.pop()
                self.on_stack[node] = False
                scc.append(node)
                if node == at:
                    break
            self.sccs.append(scc)

    def find_sccs(self):
        for i in range(self.n):
            if self.ids[i] == -1:
                self._dfs(i)
        return self.sccs

# --- CI/CD Automated Test ---
if __name__ == '__main__':
    # Graph: 0->1, 1->2, 2->0 (Cluster 1) | 3->4 (Cluster 2)
    graph = [[1], [2], [0], [4], []]
    tarjan = TarjanSCC(5, graph)
    sccs = tarjan.find_sccs()
    
    assert len(sccs) == 3 # {0,1,2}, {3}, {4}
    print("Python Tarjan SCC Test Passed!")